server/src/main/java/org/elasticsearch/action/search/SearchQueryThenFetchAsyncAction.java

```diff
@@ -788,34 +778,46 @@ void onShardDone() {
                         relevantShardIndices.set(localIndex);
                     }
                 }
-                final Object[] results = new Object[queryPhaseResultConsumer.getNumShards()];
-                for (int i = 0; i < results.length; i++) {
-                    var result = queryPhaseResultConsumer.results.get(i);
-                    if (result == null) {
-                        results[i] = failures.get(i);
-                    } else {
-                        // free context id and remove it from the result right away in case we don't need it anymore
-                        if (result instanceof QuerySearchResult q
-                            && q.getContextId() != null
-                            && relevantShardIndices.get(q.getShardIndex()) == false
-                            && q.hasSuggestHits() == false
-                            && q.getRankShardResult() == null
-                            && searchRequest.searchRequest.scroll() == null
-                            && isPartOfPIT(searchRequest.searchRequest, q.getContextId()) == false) {
-                            if (dependencies.searchService.freeReaderContext(q.getContextId())) {
-                                q.clearContextId();
+                final int resultCount = queryPhaseResultConsumer.getNumShards();
+                out = dependencies.transportService.newNetworkBytesStream();
+                out.setTransportVersion(channel.getVersion());
+                try {
+                    out.writeVInt(resultCount);
+                    for (int i = 0; i < resultCount; i++) {
+                        var result = queryPhaseResultConsumer.results.get(i);
+                        if (result == null) {
+                            out.writeBoolean(false);
+                            out.writeException(failures.remove(i));
```

Comment:
This kind of trick is neatly enabled if we serialize the response inline, makes the lifecycle of the failures map entirely obvious. The same trick could be played on the individual responses but it's a little more complicated and I didn't want to muddy the waters with a tricky change to the ref-counting by doing it in here.